src/sentry/services/hybrid_cloud/user.py

```diff
@@ -0,0 +1,147 @@
+from sentry.models.organizationmember import OrganizationMember
+from sentry.models.user import User
+from sentry.services.hybrid_cloud import (
+    CreateStubFromBase,
+    InterfaceWithLifecycle,
+    silo_mode_delegation,
+)
+from sentry.silo import SiloMode
+
+
+@dataclass(frozen=True)
+class APIUser:
+    id: int = -1
+    pk: int = -1
+    name: str = ""
+    email: str = ""
+    username: str = ""
+    actor_id: int = -1
+
+    # TODO: Extract these "view" methods to BaseUser class?
+    # BaseUser class would also be useful for catching `isinstance(foo, User)` checks
+    # which may no longer mean what they used to mean
+    def class_name(self):
+        """Necessary for integrations that introspect on user objects"""
+        return "User"
+
+    def get_display_name(self):
+        return self.name or self.email or self.username
+
+    # TODO: Extract to base dataclass?
+    @classmethod
+    def from_orm(cls, orm: Any):
+        # fetch the constructor's signature
+        cls_fields = {field for field in signature(cls).parameters}
+
+        # isolate params we care about
+        args = {}
+        for name, val in orm.__dict__.items():
+            if name in cls_fields:
+                args[name] = val
+
+        args["pk"] = orm.pk
+        ret = cls(**args)
+        return ret
+
+
+class UserService(InterfaceWithLifecycle):
+    @abstractmethod
+    def get_many_by_email(self, email: str, is_active=True) -> List[APIUser]:
+        """
+        Return a list of active users with verified emails matching the parameter
+        :param email:
+        A case insensitive email to match
+        :return:
+        """
+        pass
+
+    @abstractmethod
+    def get_from_group(self, group: Group) -> List[APIUser]:
+        """Get all users in all teams in a given Group's project."""
+        pass
+
+    @abstractmethod
+    def get_many(self, user_ids: Iterable[int], is_active=True) -> List[APIUser]:
+        """
+        This method returns User objects given an iterable of IDs
+        :param user_ids:
+        A list of user IDs to fetch
+        :return:
+        """
+        pass
+
+    def get_by_actor_id(self, actor_id: int) -> Optional[APIUser]:
+        """
+        This method returns a User object given an actor ID
+        :param actor_id:
+        An actor ID to fetch
+        :return:
+        """
+        pass
+
+    def get_user(self, user_id: int, is_active=True) -> Optional[APIUser]:
+        """
+        This method returns a User object given an ID
+        :param user_id:
+        A user ID to fetch
+        :return:
+        """
+        users = self.get_many([user_id], is_active=is_active)
+        if len(users) > 0:
+            return users[0]
+        else:
+            return None
+
+    # TODO: Extract to base service?
+    def to_api(self, resp: Union[Sequence[User], Optional[User]]) -> APIUser:
+        if resp is None:
+            return None
+        if type(resp) is APIUser:
```

Comment:
Why is this conditional possible?  The typing suggests it should not be.

Reply:
Good catch. The typing was wrong.